app/models/partnership_csv_upload.rb

```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class PartnershipCsvUpload < ApplicationRecord
+  has_one_attached :csv
+  # does this need a delivery partner reference ?
+  belongs_to :lead_provider, optional: true
+
+  validate :csv_validation
+
+  MAX_FILE_SIZE = 2.megabytes
+
+private
+
+  def csv_validation
+    return unless csv.attached?
```

Comment:
Could you not add the "Please select a CSV file to upload." here? That way, the controller becomes 
```
@partnership_csv_upload = PartnershipCsvUpload.new(upload_params.merge(lead_provider_id: current_user.id))
render :new and return unless @partnership_csv_upload.valid?
```
or similar

Reply:
I looked at that, validating on the attachement is different to an normal attribute, see the comment re the optional dependency on lead_provider.